ddtrace/tracer/metrics.go

```diff
@@ -0,0 +1,98 @@
+package tracer
+
+import (
+	"runtime"
+	"runtime/debug"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+// defaultMetricsReportInterval specifies the interval at which runtime metrics will
+// be reported.
+const defaultMetricsReportInterval = 10 * time.Second
+
+type gauger interface {
+	Gauge(name string, value float64, tags []string, rate float64) error
+}
+
+// reportMetrics periodically reports go runtime metrics to the specified gauger at
+// the given interval.
+func (t *tracer) reportMetrics(statsd gauger, interval time.Duration) {
+	var (
+		ms   runtime.MemStats
+		tags []string
+	)
+	if t.config.serviceName != "" {
+		tags = append(tags, "service:"+t.config.serviceName)
+	}
+	if t.config.hostname != "" {
+		tags = append(tags, "host:"+t.config.hostname)
+	}
+	if v, ok := t.config.globalTags[ext.Environment]; ok {
+		if vv, ok := v.(string); ok {
+			tags = append(tags, "env:"+vv)
+		}
+	}
+	gc := debug.GCStats{PauseQuantiles: make([]time.Duration, 5)}
+
+	tick := time.NewTicker(interval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+			log.Debug("Reporting runtime metrics...")
+			runtime.ReadMemStats(&ms)
+			debug.ReadGCStats(&gc)
+
```

Comment:
Some things below should be counters instead of gauges.
It depends whether they increment or report a current measurement.
Eg: `num_cgo_call`, `mem_stats.lookups`, `mem_stats.mallocs` etc.

Reply:
I’m not sure if this is a question or a statement. To me it seemed like all of them are gauges: reporting the total since process start.

Reply:
Yep, confirmed. The docs say it's the "cumulative" count. It doesn't make sense for them to reset between `ReadMemStats` calls. That would just mess with potential concurrent callers. Am I missing something?

```go
// ReadMemStats populates m with memory allocator statistics.
//
// The returned memory allocator statistics are up to date as of the
// call to ReadMemStats. This is in contrast with a heap profile,
// which is a snapshot as of the most recently completed garbage
// collection cycle.
```

Reply:
Thanks for correcting me on this (and sorry for any time lost confirming). I learned a few things!